src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -691,13 +688,33 @@ private void PopulateSessionProperties(OpenIdConnectMessage message, Authenticat
         protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCodeAsync(OpenIdConnectMessage tokenEndpointRequest)
         {
             Logger.RedeemingCodeForTokens();
+
             var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
             requestMessage.Content = new FormUrlEncodedContent(tokenEndpointRequest.Parameters);
+
             var responseMessage = await Backchannel.SendAsync(requestMessage);
-            responseMessage.EnsureSuccessStatusCode();
-            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
-            var jsonTokenResponse = JObject.Parse(tokenResonse);
-            return new OpenIdConnectMessage(jsonTokenResponse);
+
+            // Error handling:
+            // 1. If the response body is json, convert it to an OpenIdConnectMessage.
+            // 2. If the response's status code is not in 2XX range, throw OpenIdConnectProtocolException. If the body is correct parsed, 
+            //    pass the error information from body to the exception.
+            // 3. If the response body is not json, throw OpenIdConnectProtocolException.
+            if (string.Equals(responseMessage.Content.Headers.ContentType.MediaType, "applicaiton/json", StringComparison.OrdinalIgnoreCase))
```

Comment:
Also, won't this line null-ref if no `Content-Type` header is returned?

Related remark: https://github.com/aspnet/Security/issues/985#issuecomment-249035900